File: python/flexaidds/dift.py
"""DiFT — Discrete Fourier Transform torsional parametrization for FlexAID∆S.

Pure-Python (NumPy) mirror of the C++ ``dift`` engine (``LIB/DiFT/``). Produces
numerically identical results to the compiled core; the C++ path exists only as
a speed option. This module is always importable — it has no dependency on the
compiled ``_core`` extension.

Method
------
A dihedral's torsional potential ``V(φ)`` and its Fourier spectrum ``{Aₙ, ωₙ}``
are conjugate representations of the same object. One FFT yields two payoffs:

* **Energy**  — a truncated cosine series ``V(φ) = mean + Σ Aₙ cos(nφ − ωₙ)``.
* **Entropy** — the same spectrum gives the 1-D torsional partition function
  ``z = ⟨exp(−βV)⟩`` and hence a rigorous per-bond torsional entropy
  ``S_tors = (⟨V⟩ − F)/T`` — the ΔS contribution an entropy-driven docking
  engine should actually use, instead of a heuristic rotatable-bond count.

Truncation is **Shannon-collapse**: the spectral entropy ``H_spec`` of the
normalized power spectrum sets the effective mode count ``N_eff = exp(H_spec)``;
the ``⌈N_eff⌉`` highest-power terms are retained. No user threshold.

Reference
---------
Flores-Trujillo, Rodríguez-Segura, Amador-Bedolla & Domínguez (2026).
"Fast Fourier Transform Enables Automated Parametrization of Complex Dihedral
Potentials in All-Atom and Coarse-Grained Force Fields."
J. Chem. Inf. Model.  DOI: 10.1021/acs.jcim.6c00123

SPDX-License-Identifier: Apache-2.0
Copyright 2026 Le Bonhomme Pharma
"""

from __future__ import annotations

import math
from dataclasses import dataclass, field
from typing import List, Sequence

import numpy as np

__all__ = [
    "kB_kcal",
    "FourierTerm",
    "TorsionalPotential",
    "TorsionalThermo",
    "DiFTEngine",
    "spectral_entropy",
    "RotatableBondTorsion",
    "TorsionalScore",
    "score_torsional",
    "make_bond_torsion",
]

# Physical constant — matches statmech / the C++ dift namespace.
kB_kcal = 0.001987206  # kcal mol⁻¹ K⁻¹

_TWO_PI = 2.0 * math.pi


def _wrap_pi(a: float) -> float:
    """Wrap an angle into (−π, π]."""
    a = math.fmod(a + math.pi, _TWO_PI)
    if a < 0.0:
        a += _TWO_PI
    return a - math.pi


# ─────────────────────────────────────────────────────────────────────────────
# Data structures
# ─────────────────────────────────────────────────────────────────────────────

@dataclass
class FourierTerm:
    """A single cosine term: ``amplitude · cos(multiplicity·φ − phase)``."""

    multiplicity: int = 0      # n — integer frequency
    amplitude: float = 0.0     # Aₙ — force constant (kcal/mol), ≥ 0
    phase: float = 0.0         # ωₙ — phase shift (radians), in (−π, π]
    power: float = 0.0         # Aₙ² — spectral power

    def __repr__(self) -> str:  # pragma: no cover - cosmetic
        return (f"<FourierTerm n={self.multiplicity} "
                f"A={self.amplitude:.4g} ω={self.phase:.4g}>")


@dataclass
class TorsionalPotential:
    """Analytical torsional potential ``V(φ) = mean + Σ Aₙ cos(nφ − ωₙ)``."""

    terms: List[FourierTerm] = field(default_factory=list)
    mean: float = 0.0               # A₀ — DC offset (kcal/mol)
    v_min: float = 0.0              # global minimum of V(φ) (kcal/mol)
    n_samples: int = 0              # M — input grid resolution
    r_squared: float = 0.0          # fit quality vs the target profile
    spectral_entropy: float = 0.0   # H_spec (nats) of the FULL spectrum
    effective_modes: float = 0.0    # N_eff = exp(H_spec)
    refinement_iters: int = 0       # QM–MM refinement iterations used

    def evaluate(self, phi: float) -> float:
        """Absolute potential at angle ``phi`` (radians)."""
        v = self.mean
        for t in self.terms:
            v += t.amplitude * math.cos(t.multiplicity * phi - t.phase)
        return v

    def relative(self, phi: float) -> float:
        """Potential relative to its global minimum — the form used to score."""
        return self.evaluate(phi) - self.v_min

    def sample(self, n: int) -> np.ndarray:
        """Sample the analytical potential on ``n`` points over [0, 2π)."""
        if n < 1:
            return np.empty(0, dtype=float)
        phi = np.arange(n, dtype=float) * (_TWO_PI / n)
        out = np.full(n, self.mean, dtype=float)
        for t in self.terms:
            out += t.amplitude * np.cos(t.multiplicity * phi - t.phase)
        return out

    @property
    def n_terms(self) -> int:
        return len(self.terms)


@dataclass
class TorsionalThermo:
    """Per-bond torsional thermodynamics (excess, vs a free rotor).

    A free rotor (``V ≡ 0``) gives ``entropy = 0``; a confined bond gives
    ``entropy < 0`` (an entropy loss) and ``minus_TS > 0`` (a ΔG penalty).
    """

    temperature: float = 300.0
    partition_function: float = 1.0   # z = ⟨exp(−βV)⟩
    free_energy: float = 0.0          # F = −kT ln z      (kcal/mol)
    mean_energy: float = 0.0          # ⟨V⟩               (kcal/mol)
    entropy: float = 0.0              # S_tors            (kcal mol⁻¹ K⁻¹)
    minus_TS: float = 0.0             # −T·S_tors         (kcal/mol)


# ─────────────────────────────────────────────────────────────────────────────
# Spectral Shannon entropy
# ─────────────────────────────────────────────────────────────────────────────

def spectral_entropy(spectrum: Sequence[FourierTerm]) -> float:
    """Shannon entropy ``H_spec`` (nats) of a power spectrum.

    Normalized powers ``pₙ = Aₙ²/ΣAₘ²`` form a distribution; ``H = −Σ pₙ ln pₙ``.
    ``exp(H)`` is the effective number of contributing frequencies.
    """
    total = sum(t.power for t in spectrum)
    if total <= 0.0:
        return 0.0
    h = 0.0
    for t in spectrum:
        p = t.power / total
        if p > 0.0:
            h -= p * math.log(p)
    return h


# ─────────────────────────────────────────────────────────────────────────────
# Core engine
# ─────────────────────────────────────────────────────────────────────────────

class DiFTEngine:
    """DiFT torsional parametrization engine (pure-Python / NumPy)."""

    def __init__(self, temperature_K: float = 300.0) -> None:
        self.set_temperature(temperature_K)

    def set_temperature(self, t_k: float) -> None:
        self._t = t_k if t_k > 0.0 else 300.0
        self._beta = 1.0 / (kB_kcal * self._t)

    @property
    def temperature(self) -> float:
        return self._t

    # ── forward transform ────────────────────────────────────────────────────
    def transform(self, profile: Sequence[float]) -> tuple[List[FourierTerm], float]:
        """Forward DiFT of an M-point torsional profile over [0, 2π).

        Returns ``(spectrum, mean)`` where ``spectrum`` lists terms n = 1…⌊M/2⌋.
        """
        x = np.asarray(profile, dtype=float)
        m = x.size
        if m < 2:
            raise ValueError("DiFT.transform: profile needs ≥ 2 samples")

        # rfft → X_0 … X_{M//2}; sign convention X_n = Σ x_k exp(−i2πnk/M).
        x_full = np.fft.rfft(x)
        inv_m = 1.0 / m
        mean = x_full[0].real * inv_m

        n_max = m // 2
        spectrum: List[FourierTerm] = []
        for n in range(1, n_max + 1):
            nyquist = (m % 2 == 0) and (n == n_max)
            scale = inv_m if nyquist else (2.0 * inv_m)
            amp = scale * abs(x_full[n])
            phase = -math.atan2(x_full[n].imag, x_full[n].real)
            if amp < 0.0:
                amp, phase = -amp, _wrap_pi(phase + math.pi)
            else:
                phase = _wrap_pi(phase)
            spectrum.append(FourierTerm(multiplicity=n, amplitude=amp,
                                        phase=phase, power=amp * amp))
        return spectrum, mean

    # ── parametrize with Shannon-collapse truncation ─────────────────────────
    def parametrize(self, profile: Sequence[float],
                    max_multiplicity: int = 0) -> TorsionalPotential:
        """Forward transform + Shannon-collapse spectral truncation."""
        spectrum, mean = self.transform(profile)

        h_spec = spectral_entropy(spectrum)
        n_eff = math.exp(h_spec)

        if max_multiplicity > 0:
            spectrum = [t for t in spectrum if t.multiplicity <= max_multiplicity]

        keep = max(1, min(len(spectrum), math.ceil(n_eff)))
        spectrum.sort(key=lambda t: t.power, reverse=True)
        spectrum = spectrum[:keep]
        spectrum.sort(key=lambda t: t.multiplicity)

        pot = TorsionalPotential(
            terms=spectrum,
            mean=mean,
            n_samples=len(profile),
            spectral_entropy=h_spec,
            effective_modes=n_eff,
        )
        model = pot.sample(len(profile))
        pot.r_squared = self.r_squared(np.asarray(profile, dtype=float), model)
        pot.v_min = float(np.min(pot.sample(720)))
        return pot

    # ── iterative QM–MM refinement (paper eq. 18) ────────────────────────────
    def refine(self, qm: Sequence[float], mm_initial: Sequence[float],
               lambda_: float = 0.5, r2_target: float = 0.98,
               max_iter: int = 50, max_multiplicity: int = 6) -> TorsionalPotential:
        """Damped, FFT band-limited refinement: ``V_{i+1} = V_i + λ·D_i``."""
        qm_a = np.asarray(qm, dtype=float)
        mm_a = np.asarray(mm_initial, dtype=float)
        m = qm_a.size
        if m < 2 or mm_a.size != m:
            raise ValueError("DiFT.refine: qm and mm_initial must share a "
                             "grid of ≥ 2 samples")

        correction = np.zeros(m, dtype=float)
        pot = TorsionalPotential()
        iters = 0
        r2 = 0.0
        for it in range(1, max_iter + 1):
            iters = it
            mm_current = mm_a + correction
            correction = correction + lambda_ * (qm_a - mm_current)
            pot = self.parametrize(correction, max_multiplicity)
            correction = pot.sample(m)
            r2 = self.r_squared(qm_a, mm_a + correction)
            if r2 >= r2_target:
                break

        pot.r_squared = r2
        pot.refinement_iters = iters
        return pot

    # ── Boltzmann inversion of a CG dihedral histogram (paper eq. 20) ────────
    def boltzmann_invert(self, histogram: Sequence[float]) -> np.ndarray:
        """Invert a dihedral-angle histogram into an energy profile.

        ``E(φ) = −kT ln p(φ)``, shifted so the minimum is 0. No Jacobian:
        dihedral angles are uniformly distributed for a free rotor.
        """
        h = np.asarray(histogram, dtype=float)
        if h.size < 2:
            raise ValueError("DiFT.boltzmann_invert: need ≥ 2 bins")
        if np.any(h < 0.0):
            raise ValueError("DiFT.boltzmann_invert: negative count")
        total = float(h.sum())
        if total <= 0.0:
            raise ValueError("DiFT.boltzmann_invert: empty histogram")

        kt = kB_kcal * self._t
        energy = np.full(h.size, np.nan, dtype=float)
        nonzero = h > 0.0
        energy[nonzero] = -kt * np.log(h[nonzero] / total)
        e_max = float(np.nanmax(energy))
        energy[~nonzero] = e_max          # empty bins → capped well wall
        energy -= float(np.nanmin(energy))  # shift minimum to 0
        return energy

    # ── per-bond torsional thermodynamics ────────────────────────────────────
    def thermodynamics(self, pot: TorsionalPotential) -> TorsionalThermo:
        """Excess torsional thermodynamics from a parametrized potential."""
        n_fine = 1440
        phi = np.arange(n_fine, dtype=float) * (_TWO_PI / n_fine)
        v = np.full(n_fine, pot.mean, dtype=float)
        for t in pot.terms:
            v += t.amplitude * np.cos(t.multiplicity * phi - t.phase)
        v -= pot.v_min                       # energies relative to the minimum

        w = np.exp(-self._beta * v)
        z_sum = float(w.sum())
        z = z_sum / n_fine                   # ⟨exp(−βV)⟩; 1 for a free rotor
        mean_e = float((w * v).sum() / z_sum) if z_sum > 0.0 else 0.0
        f = -kB_kcal * self._t * math.log(z)

        th = TorsionalThermo(temperature=self._t, partition_function=z,
                             free_energy=f, mean_energy=mean_e)
        th.entropy = (mean_e - f) / self._t
        th.minus_TS = -self._t * th.entropy
        return th

    # ── helpers ──────────────────────────────────────────────────────────────
    @staticmethod
    def circular_mean(angles: Sequence[float]) -> float:
        """Directional mean of phase offsets — ``atan2(Σ sin, Σ cos)``."""
        a = np.asarray(angles, dtype=float)
        s = float(np.sin(a).sum())
        c = float(np.cos(a).sum())
        if s == 0.0 and c == 0.0:
            return 0.0
        return math.atan2(s, c)

    @staticmethod
    def r_squared(observed: np.ndarray, model: np.ndarray) -> float:
        """Coefficient of determination R² between two equal-length profiles."""
        obs = np.asarray(observed, dtype=float)
        mod = np.asarray(model, dtype=float)
        if obs.size == 0 or mod.size != obs.size:
            return 0.0
        ss_res = float(np.sum((obs - mod) ** 2))
        ss_tot = float(np.sum((obs - obs.mean()) ** 2))
        if ss_tot <= 0.0:
            return 1.0 if ss_res <= 1e-18 else 0.0
        return 1.0 - ss_res / ss_tot


# ─────────────────────────────────────────────────────────────────────────────
# GA scoring adapter — mirrors LIB/DiFT/DiFTGAAdapter.h
# ─────────────────────────────────────────────────────────────────────────────

@dataclass
class RotatableBondTorsion:
    """A ligand rotatable bond and its DiFT-parametrized potential."""

    potential: TorsionalPotential
    gene_index: int = -1


@dataclass
class TorsionalScore:
    """Decomposed torsional contribution to the binding free energy."""

    energy: float = 0.0     # Σ V_tors,b relative to each well minimum
    minus_TS: float = 0.0   # Σ −T·S_tors,b — torsional confinement penalty
    n_bonds: int = 0

    def total(self) -> float:
        """Free-energy contribution the GA adds to the CF score."""
        return self.energy + self.minus_TS


def score_torsional(bonds: Sequence[RotatableBondTorsion],
                    dihedral_angles_rad: Sequence[float],
                    temperature_K: float = 300.0) -> TorsionalScore:
    """Score a pose's torsional state — one FFT-derived potential per bond."""
    score = TorsionalScore()
    if len(bonds) != len(dihedral_angles_rad):
        return score  # caller contract violated → zero contribution
    engine = DiFTEngine(temperature_K)
    for bond, phi in zip(bonds, dihedral_angles_rad):
        score.energy += bond.potential.relative(phi)
        score.minus_TS += engine.thermodynamics(bond.potential).minus_TS
        score.n_bonds += 1
    return score


def make_bond_torsion(profile: Sequence[float], gene_index: int,
                      temperature_K: float = 300.0,
                      max_multiplicity: int = 6) -> RotatableBondTorsion:
    """Parametrize a raw torsional profile into a :class:`RotatableBondTorsion`."""
    engine = DiFTEngine(temperature_K)
    return RotatableBondTorsion(
        potential=engine.parametrize(profile, max_multiplicity),
        gene_index=gene_index,
    )

File: LIB/DiFT/DiFT.cpp
// DiFT.cpp — Discrete Fourier Transform torsional parametrization for FlexAID∆S
//
// Copyright 2026 Le Bonhomme Pharma
// SPDX-License-Identifier: Apache-2.0
//
// Implementation of the DiFT engine. Self-contained: depends only on the
// C++ standard library (no GPL, no external FFT dependency — the radix-2
// transform is implemented here for licensing cleanliness).
//
// See DiFT.h for the method, references, and the energy/entropy rationale.

#include "DiFT.h"

#include <algorithm>
#include <cmath>
#include <complex>
#include <numeric>
#include <stdexcept>

namespace dift {

namespace {

constexpr double kTwoPi = 6.283185307179586476925286766559;

// True iff n is a power of two and n ≥ 2.
inline bool is_pow2(std::size_t n) noexcept {
    return n >= 2 && (n & (n - 1)) == 0;
}

// In-place iterative radix-2 Cooley–Tukey FFT.
// Sign convention:  X_n = Σ_k x_k · exp(−i·2π·n·k/N)   — O(N log N).
void fft_radix2(std::vector<std::complex<double>>& a) {
    const std::size_t n = a.size();
    if (n < 2) return;

    // Bit-reversal permutation.
    for (std::size_t i = 1, j = 0; i < n; ++i) {
        std::size_t bit = n >> 1;
        for (; j & bit; bit >>= 1) j ^= bit;
        j ^= bit;
        if (i < j) std::swap(a[i], a[j]);
    }

    for (std::size_t len = 2; len <= n; len <<= 1) {
        const double ang = -kTwoPi / static_cast<double>(len);
        const std::complex<double> wlen(std::cos(ang), std::sin(ang));
        for (std::size_t i = 0; i < n; i += len) {
            std::complex<double> w(1.0, 0.0);
            for (std::size_t k = 0; k < (len >> 1); ++k) {
                const std::complex<double> u = a[i + k];
                const std::complex<double> v = a[i + k + (len >> 1)] * w;
                a[i + k]               = u + v;
                a[i + k + (len >> 1)]  = u - v;
                w *= wlen;
            }
        }
    }
}

// Exact direct DFT for arbitrary (non-power-of-two) lengths — O(N²).
// Torsional grids are small (typically 24–72 points), so this is negligible
// and keeps the transform mathematically exact rather than zero-padding a
// periodic signal (which would distort its spectrum).
std::vector<std::complex<double>> dft_direct(std::span<const double> x) {
    const std::size_t M = x.size();
    std::vector<std::complex<double>> X(M);
    for (std::size_t n = 0; n < M; ++n) {
        std::complex<double> s(0.0, 0.0);
        for (std::size_t k = 0; k < M; ++k) {
            const double ang = -kTwoPi * static_cast<double>(n) *
                               static_cast<double>(k) / static_cast<double>(M);
            s += x[k] * std::complex<double>(std::cos(ang), std::sin(ang));
        }
        X[n] = s;
    }
    return X;
}

// Wrap an angle into (−π, π].
inline double wrap_pi(double a) noexcept {
    a = std::fmod(a + M_PI, kTwoPi);
    if (a < 0.0) a += kTwoPi;
    return a - M_PI;
}

} // namespace

// ─────────────────────────────────────────────────────────────────────────────
DiFTEngine::DiFTEngine(double temperature_K) {
    set_temperature(temperature_K);
}

void DiFTEngine::set_temperature(double T_K) noexcept {
    T_    = (T_K > 0.0) ? T_K : 300.0;
    beta_ = 1.0 / (kB_kcal * T_);
}

// ─────────────────────────────────────────────────────────────────────────────
std::vector<FourierTerm>
DiFTEngine::transform(std::span<const double> profile, double& mean_out) const {
    const std::size_t M = profile.size();
    if (M < 2)
        throw std::invalid_argument("DiFT::transform: profile needs ≥ 2 samples");
    for (double v : profile) {
        if (!std::isfinite(v))
            throw std::invalid_argument("DiFT::transform: profile contains non-finite values");
    }

    // Forward transform → full complex spectrum X_0 … X_{M−1}.
    std::vector<std::complex<double>> X;
    if (is_pow2(M)) {
        X.resize(M);
        for (std::size_t k = 0; k < M; ++k) X[k] = std::complex<double>(profile[k], 0.0);
        fft_radix2(X);
    } else {
        X = dft_direct(profile);
    }

    const double invM = 1.0 / static_cast<double>(M);

    // DC component → mean offset.
    mean_out = X[0].real() * invM;

    // Real-signal reconstruction:
    //   x_k = mean + Σ_{n=1}^{M/2} Aₙ·cos(n·φ_k − ωₙ),  φ_k = 2πk/M
    // with, for 1 ≤ n < M/2:  Aₙ = (2/M)|Xₙ|,  ωₙ = −arg(Xₙ).
    // The Nyquist term (n = M/2, M even) carries weight 1/M.
    const std::size_t n_max = M / 2;
    std::vector<FourierTerm> spectrum;
    spectrum.reserve(n_max);

    for (std::size_t n = 1; n <= n_max; ++n) {
        const bool   nyquist = (M % 2 == 0) && (n == n_max);
        const double scale   = nyquist ? invM : (2.0 * invM);
        double amp           = scale * std::abs(X[n]);
        double phase         = -std::atan2(X[n].imag(), X[n].real());

        // Normalize to amplitude ≥ 0 (fold a negative sign into a π phase flip).
        if (amp < 0.0) { amp = -amp; phase = wrap_pi(phase + M_PI); }
        else            { phase = wrap_pi(phase); }

        FourierTerm t;
        t.multiplicity = static_cast<int>(n);
        t.amplitude    = amp;
        t.phase        = phase;
        t.power        = amp * amp;
        spectrum.push_back(t);
    }
    return spectrum;
}

// ─────────────────────────────────────────────────────────────────────────────
TorsionalPotential
DiFTEngine::parametrize(std::span<const double> profile, int max_multiplicity) const {
    double mean = 0.0;
    std::vector<FourierTerm> spectrum = transform(profile, mean);

    // Spectral Shannon entropy of the FULL spectrum (before any truncation).
    const double H_spec = spectral_entropy(spectrum);
    const double N_eff  = std::exp(H_spec);

    // Anti-overfit guard: drop frequencies above the requested cap.
    if (max_multiplicity > 0) {
        std::erase_if(spectrum, [max_multiplicity](const FourierTerm& t) {
            return t.multiplicity > max_multiplicity;
        });
    }

    // Shannon-collapse truncation: keep the ⌈N_eff⌉ highest-power terms.
    // The spectrum's own entropy decides the term count — no user threshold.
    std::size_t keep = static_cast<std::size_t>(std::ceil(N_eff));
    keep = std::clamp<std::size_t>(keep, 1, spectrum.size());

    std::sort(spectrum.begin(), spectrum.end(),
              [](const FourierTerm& a, const FourierTerm& b) {
                  return a.power > b.power;
              });
    if (spectrum.size() > keep) spectrum.resize(keep);

    // Retained terms back in ascending-multiplicity order (canonical form).
    std::sort(spectrum.begin(), spectrum.end(),
              [](const FourierTerm& a, const FourierTerm& b) {
                  return a.multiplicity < b.multiplicity;
              });

    TorsionalPotential pot;
    pot.terms            = std::move(spectrum);
    pot.mean             = mean;
    pot.n_samples        = static_cast<int>(profile.size());
    pot.spectral_entropy = H_spec;
    pot.effective_modes  = N_eff;
    pot.refinement_iters = 0;

    // Fit quality of the truncated model vs the input profile.
    std::vector<double> model = pot.sample(static_cast<int>(profile.size()));
    pot.r_squared = r_squared(profile, model);

    // Cache the global minimum (used for relative energies / scoring).
    std::vector<double> fine = pot.sample(720);
    pot.v_min = *std::min_element(fine.begin(), fine.end());

    return pot;
}

// ─────────────────────────────────────────────────────────────────────────────
TorsionalPotential
DiFTEngine::refine(std::span<const double> qm,
                   std::span<const double> mm_initial,
                   double lambda, double r2_target,
                   int max_iter, int max_multiplicity) const {
    const std::size_t M = qm.size();
    if (M < 2 || mm_initial.size() != M)
        throw std::invalid_argument(
            "DiFT::refine: qm and mm_initial must share a grid of ≥ 2 samples");

    // The torsional correction satisfies  mm_initial + V_t ≈ qm.
    // We approach it through the damped, FFT band-limited loop of eq. 18.
    std::vector<double> correction(M, 0.0);
    std::vector<double> mm_current(M, 0.0);
    TorsionalPotential  pot;
    int    iters = 0;
    double r2    = 0.0;

    for (int it = 1; it <= max_iter; ++it) {
        iters = it;
        // D_i = V_QM − V_MM,i
        for (std::size_t k = 0; k < M; ++k)
            mm_current[k] = mm_initial[k] + correction[k];

        // V_{i+1} = V_i + λ·D_i
        for (std::size_t k = 0; k < M; ++k)
            correction[k] += lambda * (qm[k] - mm_current[k]);

        // Band-limit the correction through the FFT (paper: FFT applied to V_i).
        pot        = parametrize(correction, max_multiplicity);
        correction = pot.sample(static_cast<int>(M));

        // Convergence: does mm_initial + correction reproduce qm?
        for (std::size_t k = 0; k < M; ++k)
            mm_current[k] = mm_initial[k] + correction[k];
        r2 = r_squared(qm, mm_current);
        if (r2 >= r2_target) break;
    }

    pot.r_squared        = r2;
    pot.refinement_iters = iters;
    return pot;
}

// ─────────────────────────────────────────────────────────────────────────────
std::vector<double>
DiFTEngine::boltzmann_invert(std::span<const double> histogram) const {
    const std::size_t M = histogram.size();
    if (M < 2)
        throw std::invalid_argument("DiFT::boltzmann_invert: need ≥ 2 bins");

    double total = 0.0;
    for (double c : histogram) {
        if (!std::isfinite(c))
            throw std::invalid_argument("DiFT::boltzmann_invert: non-finite count");
        if (c < 0.0)
            throw std::invalid_argument("DiFT::boltzmann_invert: negative count");
        total += c;
    }
    if (total <= 0.0)
        throw std::invalid_argument("DiFT::boltzmann_invert: empty histogram");

    // E(φ) = −kT ln p(φ). No Jacobian: dihedral angles are uniformly
    // distributed for a free rotor (unlike bond lengths or bond angles).
    const double kT = kB_kcal * T_;
    std::vector<double> energy(M);
    double e_min =  1e300;
    double e_max = -1e300;
    for (std::size_t k = 0; k < M; ++k) {
        if (histogram[k] > 0.0) {
            energy[k] = -kT * std::log(histogram[k] / total);
            e_min = std::min(e_min, energy[k]);
            e_max = std::max(e_max, energy[k]);
        } else {
            energy[k] = std::nan("");   // flag empty bins; filled below
        }
    }
    // Empty bins → capped well wall at the highest observed energy.
    for (double& e : energy)
        if (std::isnan(e)) e = e_max;
    // Shift so the global minimum sits at 0 (relative energies).
    for (double& e : energy) e -= e_min;
    return energy;
}

// ─────────────────────────────────────────────────────────────────────────────
TorsionalThermo
DiFTEngine::thermodynamics(const TorsionalPotential& pot) const {
    // Integrate the 1-D torsional partition function on a fine grid using the
    // analytical potential. Energies are taken relative to the global minimum
    // so the result is the EXCESS thermodynamics vs an (unconfined) free rotor.
    constexpr int Nf = 1440;
    double Z = 0.0, EZ = 0.0;
    for (int j = 0; j < Nf; ++j) {
        const double phi = kTwoPi * static_cast<double>(j) / static_cast<double>(Nf);
        const double V   = pot.evaluate(phi) - pot.v_min;   // ≥ 0
        const double w   = std::exp(-beta_ * V);
        Z  += w;
        EZ += w * V;
    }
    const double invNf = 1.0 / static_cast<double>(Nf);
    const double z     = Z * invNf;                 // ⟨exp(−βV)⟩; 1 for a free rotor
    const double mean  = (Z > 0.0) ? (EZ / Z) : 0.0;
    const double F     = -kB_kcal * T_ * std::log(z);

    TorsionalThermo th;
    th.temperature        = T_;
    th.partition_function = z;
    th.free_energy        = F;
    th.mean_energy        = mean;
    th.entropy            = (mean - F) / T_;        // ≤ 0: confinement = entropy loss
    th.minus_TS           = -T_ * th.entropy;       // ≥ 0: the ΔG penalty
    return th;
}

// ─────────────────────────────────────────────────────────────────────────────
double DiFTEngine::circular_mean(std::span<const double> angles) noexcept {
    double s = 0.0, c = 0.0;
    for (double a : angles) { s += std::sin(a); c += std::cos(a); }
    if (s == 0.0 && c == 0.0) return 0.0;
    return std::atan2(s, c);
}

double DiFTEngine::r_squared(std::span<const double> observed,
                             std::span<const double> model) noexcept {
    const std::size_t n = observed.size();
    if (n == 0 || model.size() != n) return 0.0;

    double mean = 0.0;
    for (double v : observed) mean += v;
    mean /= static_cast<double>(n);

    double ss_res = 0.0, ss_tot = 0.0;
    for (std::size_t i = 0; i < n; ++i) {
        const double d_res = observed[i] - model[i];
        const double d_tot = observed[i] - mean;
        ss_res += d_res * d_res;
        ss_tot += d_tot * d_tot;
    }
    if (ss_tot <= 0.0)                     // flat target
        return (ss_res <= 1e-18) ? 1.0 : 0.0;
    return 1.0 - ss_res / ss_tot;
}

// ─────────────────────────────────────────────────────────────────────────────
double TorsionalPotential::evaluate(double phi) const noexcept {
    double v = mean;
    for (const FourierTerm& t : terms)
        v += t.amplitude *
             std::cos(static_cast<double>(t.multiplicity) * phi - t.phase);
    return v;
}

std::vector<double> TorsionalPotential::sample(int n) const {
    std::vector<double> out;
    if (n < 1) return out;
    out.reserve(static_cast<std::size_t>(n));
    for (int j = 0; j < n; ++j)
        out.push_back(evaluate(kTwoPi * static_cast<double>(j) /
                                static_cast<double>(n)));
    return out;
}

// ─────────────────────────────────────────────────────────────────────────────
double spectral_entropy(std::span<const FourierTerm> spectrum) noexcept {
    double total = 0.0;
    for (const FourierTerm& t : spectrum) total += t.power;
    if (total <= 0.0) return 0.0;

    double H = 0.0;
    for (const FourierTerm& t : spectrum) {
        const double p = t.power / total;
        if (p > 0.0) H -= p * std::log(p);
    }
    return H;
}

} // namespace dift
